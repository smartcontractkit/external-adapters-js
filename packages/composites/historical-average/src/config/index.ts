import * as CoinMarketCap from '@chainlink/coinmarketcap-adapter'
import { Requester, util } from '@chainlink/ea-bootstrap'
import { AdapterImplementation, Config as DefaultConfig } from '@chainlink/types'

export type SourceRequestOptions = { [source: string]: DefaultConfig }
export const NAME = 'HISTORICAL_AVERAGE'

export type Config = DefaultConfig & {
  sources: SourceRequestOptions
  defaultSource?: string
}

export const adapters: AdapterImplementation[] = [CoinMarketCap]

export type Source = typeof adapters[number]['NAME']

export const makeConfig = (prefix = ''): Config => {
  const sources: SourceRequestOptions = {}

  for (const a of adapters) {
    const name = a.NAME
    const url = util.getURL(name.toUpperCase())
    if (url) {
      const defaultConfig = Requester.getDefaultConfig(prefix)
      defaultConfig.api.baseURL = url
      defaultConfig.api.method = 'post'
      sources[name.toLowerCase()] = defaultConfig
    }
  }

  return {
    sources,
    defaultSource: util.getEnv('DEFAULT_SOURCE'),
  }
}
