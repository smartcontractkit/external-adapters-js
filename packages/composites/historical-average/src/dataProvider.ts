import { AxiosRequestConfig, AdapterResponse } from '@chainlink/ea-bootstrap'
import { AdapterConfigError, AdapterInputError, Logger, Requester } from '@chainlink/ea-bootstrap'
import { adapter as cmc } from '@chainlink/coinmarketcap-adapter'
import { ResponseSchema } from '@chainlink/coinmarketcap-adapter/src/endpoint/historical'

export type ResponsePayload = {
  timestamp: Date
  price: number
}[]

export const getPriceProvider =
  (source: string, jobRunID: string, apiConfig: AxiosRequestConfig) =>
  async (
    base: string,
    quote: string,
    fromDate: Date,
    toDate: Date,
    interval: string,
  ): Promise<ResponsePayload> => {
    try {
      switch (source.toUpperCase()) {
        case cmc.name:
          return getCoinMarketCapPrice(jobRunID, base, quote, fromDate, toDate, interval, apiConfig)
      }
    } catch (error) {
      Logger.error(`Request to ${source} adapter failed: ${error}`)
      throw new AdapterConfigError({
        jobRunID,
        message: `Failed to request the ${source} adapter. Ensure that the ${source} adapter is working correctly & ${source.toUpperCase()}_ADAPTER_URL environment variable is correctly pointed to the adapter location.`,
      })
    }

    throw new AdapterInputError({
      jobRunID,
      message: `No historical data implementation for source ${source}!`,
    })
  }

const getCoinMarketCapPrice = async (
  jobRunID: string,
  base: string,
  quote: string,
  fromDate: Date,
  toDate: Date,
  interval: string,
  config: AxiosRequestConfig,
): Promise<ResponsePayload> => {
  const data = {
    id: jobRunID,
    data: {
      endpoint: 'historical',
      base,
      quote,
      start: fromDate.toISOString(),
      end: toDate.toISOString(),
      interval,
      // Specifying count is not necessary since we have provided a [start, end] interval
    },
  }
  const response = await Requester.request<AdapterResponse>({ ...config, data })
  const responseData = response.data as unknown as ResponseSchema
  return responseData.data.quotes.map((entry) => ({
    timestamp: new Date(entry.timestamp),
    price: entry.quote[quote.toUpperCase()].price,
  }))
}
