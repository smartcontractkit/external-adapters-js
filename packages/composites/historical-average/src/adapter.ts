import {
  AdapterError,
  AdapterConfigError,
  AdapterInputError,
  Requester,
  Validator,
} from '@chainlink/ea-bootstrap'
import {
  AdapterResponse,
  ExecuteWithConfig,
  ExecuteFactory,
  InputParameters,
} from '@chainlink/ea-bootstrap'
import { getPriceProvider } from './dataProvider'
import { Config, makeConfig } from './config'
import { Decimal } from 'decimal.js'

Decimal.set({ precision: 100 })

export type TInputParameters = {
  from: string
  to: string
  source?: string
  fromDate?: string
  toDate?: string
  days?: string
  interval?: string
}

const inputParameters: InputParameters<TInputParameters> = {
  from: {
    required: true,
    aliases: ['base', 'coin'],
  },
  to: {
    required: true,
    aliases: ['quote', 'market'],
  },
  source: {
    required: false,
  },
  fromDate: {
    required: false,
  },
  toDate: {
    required: false,
  },
  days: {
    required: false,
  },
  interval: {
    required: false,
  },
}

export const execute: ExecuteWithConfig<Config, TInputParameters> = async (
  input,
  _,
  config,
): Promise<AdapterResponse> => {
  const validator = new Validator(input, inputParameters)

  const jobRunID = validator.validated.id

  const from = validator.validated.data.from
  const to = validator.validated.data.to
  const source = (validator.validated.data.source || config.defaultSource || '').toLowerCase()
  if (source === '') {
    throw new AdapterConfigError({
      statusCode: 400,
      message: 'No source provided in request or the DEFAULT_SOURCE env var!',
    })
  }
  if (!(source in config.sources)) {
    throw new AdapterConfigError({
      statusCode: 400,
      message: `The ADAPTER_URL has not been configured for source ${source}`,
    })
  }

  const fromDateInput = validator.validated.data.fromDate
  const toDateInput = validator.validated.data.toDate
  const days = validator.validated.data.days
  const interval = validator.validated.data.interval as string
  // TODO: non-nullable default types

  const { fromDate, toDate } = getFromToDates(fromDateInput, toDateInput, days)

  const apiOptions = config.sources[source].api
  if (!apiOptions)
    throw new AdapterError({
      statusCode: 400,
      message: `${source} configuration not found`,
    })
  const provider = getPriceProvider(source, jobRunID, apiOptions)
  const providerResponse = await provider(from, to, fromDate, toDate, interval)
  const result = providerResponse
    .reduce((sum, elem) => sum.add(elem.price), new Decimal(0))
    .div(providerResponse.length)

  const response = {
    data: providerResponse,
    status: 200,
    statusText: 'success',
    headers: {},
    config: {},
  }
  return Requester.success(
    jobRunID,
    Requester.withResult(response, result.toNumber()),
    config.verbose,
  )
}

const addDays = (date: Date, days: number): Date => {
  const result = new Date(date)
  result.setDate(result.getDate() + days)
  return result
}

/**
 * Get the from & to date from a set of optional params
 * @example
 * // returns (startDate, toDate)
 * getFromToDates(startDate, toDate, 5);
 * @example
 * // returns (toDate - 5 days, toDate)
 * getFromToDates(undefined, toDate, 5);
 * @example
 * // returns (startDate, startDate + 5 days)
 * getFromToDates(startDate, undefined, 5);
 * @returns { fromDate: Date, toDate: Date } Returns the from/to dates.
 */
export const getFromToDates = (
  from?: string,
  to?: string,
  days?: string | number,
): { fromDate: Date; toDate: Date } => {
  if (from && to) {
    const fromDate = new Date(from)
    const toDate = new Date(to)
    if (fromDate >= toDate) {
      throw new AdapterInputError({
        statusCode: 400,
        message: 'The "fromDate" must be before the "toDate"',
      })
    }

    return { fromDate, toDate }
  } else if (!days) {
    throw new AdapterInputError({
      statusCode: 400,
      message: 'Missing either from/to dates and days param',
    })
  }
  const numDays = Number(days)
  if (numDays <= 0) {
    throw new AdapterInputError({
      statusCode: 400,
      message: 'The days param is less than or equal to 0',
    })
  }

  if (from) {
    const fromDate = new Date(from)
    const toDate = addDays(fromDate, numDays)
    return { fromDate, toDate }
  } else if (to) {
    const toDate = new Date(to)
    const fromDate = addDays(toDate, -numDays)
    return { fromDate, toDate }
  }

  throw new AdapterInputError({
    statusCode: 400,
    message: 'Missing both from/to dates',
  })
}

export const makeExecute: ExecuteFactory<Config, TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
