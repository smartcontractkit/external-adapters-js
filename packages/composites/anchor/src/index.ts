import { expose } from '@chainlink/ea-bootstrap'
import { endpointSelector, makeExecute } from './adapter'
import { makeConfig, NAME } from './config'

/**
 * NOTE: This adapter could probably be made more general in the future.  All it is really doing is fetching an intermediary token price
 * using the Token Allocation adapter and handling the price after that.
 */

const adapterContext = { name: NAME }

const { server } = expose(adapterContext, makeExecute(), undefined, endpointSelector)

export { NAME, makeConfig, makeExecute, server }
