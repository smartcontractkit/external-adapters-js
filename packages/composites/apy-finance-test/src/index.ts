import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { allocations, tvl } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: tvl.name,
  name: 'APY_FINANCE_TEST',
  config,
  endpoints: [allocations, tvl],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
