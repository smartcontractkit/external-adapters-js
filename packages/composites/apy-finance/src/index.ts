import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'APY-Finance'

export = { NAME, makeConfig, makeExecute, ...expose(NAME, makeExecute()) }
