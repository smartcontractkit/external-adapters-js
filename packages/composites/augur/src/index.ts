import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig } from './config'

const NAME = 'AUGUR'

export = { NAME, makeConfig, makeExecute, ...expose(NAME, makeExecute()) }
