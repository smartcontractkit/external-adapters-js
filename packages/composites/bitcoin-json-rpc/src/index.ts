import { makeExecute } from './adapter'
import { expose } from '@chainlink/ea-bootstrap'

const NAME = 'BITCOIN_JSON_RPC'
export = { NAME, makeExecute, ...expose(NAME, makeExecute()) }
