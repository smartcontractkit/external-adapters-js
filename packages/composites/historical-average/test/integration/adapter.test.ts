import { AdapterRequest } from '@chainlink/ea-bootstrap'
import process from 'process'
import { server as startServer } from '../../src'
import { mockCoinmarketcapAdapter } from './fixtures'
import { setupExternalAdapterTest } from '@chainlink/ea-test-helpers'
import type { SuiteContext } from '@chainlink/ea-test-helpers'
import { SuperTest, Test } from 'supertest'

describe('execute', () => {
  const id = '1'
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const envVariables = {
    COINMARKETCAP_ADAPTER_URL: process.env.COINMARKETCAP_ADAPTER_URL || 'http://localhost:8081',
  }
  setupExternalAdapterTest(envVariables, context)
  describe('with to/from dates', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: 'ETH',
        to: 'USD',
        fromDate: '2021-11-01',
        toDate: '2021-11-08',
        source: 'coinmarketcap',
        interval: '1d',
      },
    }

    it('should return success', async () => {
      mockCoinmarketcapAdapter()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('with from date and days', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: 'ETH',
        to: 'USD',
        fromDate: '2021-11-01',
        days: 7,
        source: 'coinmarketcap',
        interval: '1d',
      },
    }

    it('should return success', async () => {
      mockCoinmarketcapAdapter()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('with to date and days', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: 'ETH',
        to: 'USD',
        toDate: '2021-11-08',
        days: 7,
        source: 'coinmarketcap',
        interval: '1d',
      },
    }

    it('should return success', async () => {
      mockCoinmarketcapAdapter()

      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
