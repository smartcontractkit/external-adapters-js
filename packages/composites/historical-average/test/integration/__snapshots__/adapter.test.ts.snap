// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute with from date and days should return success 1`] = `
{
  "data": {
    "result": 4526.04602020095,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 4526.04602020095,
  "statusCode": 200,
}
`;

exports[`execute with to date and days should return success 1`] = `
{
  "data": {
    "result": 4526.04602020095,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 4526.04602020095,
  "statusCode": 200,
}
`;

exports[`execute with to/from dates should return success 1`] = `
{
  "data": {
    "result": 4526.04602020095,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 4526.04602020095,
  "statusCode": 200,
}
`;
